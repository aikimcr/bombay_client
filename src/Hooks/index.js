export * from './useLoginTracking';
export * from './useModelCollection';
