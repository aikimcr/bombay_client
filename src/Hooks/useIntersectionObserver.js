import { useEffect, useRef } from "react";

export function useIntersectionObserver(topRef, callback, options = {}) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) return;

    if (observer.current == null) {
      observer.current = new IntersectionObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [callback, options, topRef]);

  return observer;
}

export default useIntersectionObserver;
