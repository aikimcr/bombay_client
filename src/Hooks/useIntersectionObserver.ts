import { useEffect, useRef, RefObject } from 'react';

export interface IntersectionObserverOptions {
  root?: Element | null;
  rootMargin?: string;
  threshold?: number | number[];
}

export interface UseIntersectionObserverReturn {
  current: IntersectionObserver | null;
}

export function useIntersectionObserver(
  topRef: RefObject<Element>,
  callback: IntersectionObserverCallback,
  options: IntersectionObserverOptions = {},
): UseIntersectionObserverReturn {
  const observer = useRef<IntersectionObserver | null>(null);

  useEffect(() => {
    if (topRef.current == null) {
      return;
    }

    if (observer.current == null) {
      observer.current = new IntersectionObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      if (observer.current) {
        observer.current.disconnect();
      }
    };
  }, [callback, options, topRef]);

  return observer;
}

export default useIntersectionObserver;
