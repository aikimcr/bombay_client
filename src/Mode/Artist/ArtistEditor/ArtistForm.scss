@use '@styles/mixins.scss';

.artist-form {
  @include mixins.standard-dialog;

  &__form {
    display: grid;
    grid-auto-flow: row;
    gap: 1rem;
    padding: 1rem;

    & > .controls {
      display: flex;
      justify-content: space-between;
      align-items: center;
      padding: 1rem 0;
    }
  }
}
