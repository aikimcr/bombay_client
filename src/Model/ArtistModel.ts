import ModelBase from './ModelBase';

// At this point, there isn't anything to add.
class ArtistModel extends ModelBase {}

export default ArtistModel;
