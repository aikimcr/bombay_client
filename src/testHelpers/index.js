export * from './ContextChanger';
export * from './modelTools';
