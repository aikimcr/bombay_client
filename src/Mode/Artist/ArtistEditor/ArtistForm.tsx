import React, {
  MouseEventHandler,
  ToggleEventHandler,
  useEffect,
  useRef,
  useState,
} from 'react';
import { useForm } from 'react-hook-form';
import {
  Button,
  DialogLayout,
  LabeledInput,
  ProtectedRoute,
  TextInput,
} from '../../../Components';
import { ArtistData, ArtistModel } from '../../../Model/ArtistModel';

import './ArtistForm.scss';

interface ArtistFormProps {
  artist?: ArtistModel;
  ref: ReturnType<typeof useRef<HTMLDialogElement>>;
  onClose?: (newArtist: ArtistModel) => void;
  onToggle?: ToggleEventHandler;
}

export const ArtistForm: React.FC<ArtistFormProps> = ({
  artist,
  ref,
  onClose,
  onToggle,
}): React.ReactElement => {
  const [submitDisabled, setSubmitDisabled] = useState(true);

  const { register, reset, handleSubmit, formState, setError } =
    useForm<ArtistData>({
      shouldUseNativeValidation: true,
    });

  const onSubmit = (data: ArtistData) => {
    const saveArtist = artist || new ArtistModel({});
    saveArtist.name = data.name;

    saveArtist
      .save()
      .then(() => {
        ref.current?.close();
        onClose?.call(null, saveArtist);
      })
      .catch((err: Error) => {
        console.error(err);
        setError('root.serverError', err);
      });
  };

  const handleReset = (): void => {
    setSubmitDisabled(true);
    reset({
      name: artist?.name || '',
    });
  };

  const handleClose = () => {
    handleReset();
    onClose?.call(null, artist);
  };

  const handleToggle: ToggleEventHandler = (evt) => {
    onToggle?.call(null, evt);
  };

  const closeDialog: MouseEventHandler = (evt) => {
    evt.preventDefault();
    evt.stopPropagation();
    ref.current?.close();
  };

  useEffect(() => {
    if (formState.isValid && Object.entries(formState.errors).length === 0) {
      setSubmitDisabled(false);
      return;
    }

    setSubmitDisabled(true);
  }, [formState]);

  return (
    <ProtectedRoute>
      <dialog
        ref={ref}
        className="artist-form"
        // As of "@types/react": "^19.1.13" and "@types/react-dom": "^19.1.9"
        // @ts-expect-error Typescript needs to catch up
        closedby="any"
        onClose={handleClose}
        onToggle={handleToggle}
      >
        <DialogLayout headerText="Artist Editor" closeDialog={closeDialog}>
          <form
            className="artist-form__form"
            data-testid="artist-form-form"
            onSubmit={handleSubmit(onSubmit)}
            onReset={handleReset}
          >
            <div className="info">
              <LabeledInput
                inputId="name"
                labelText="Artist Name"
                InputField={TextInput}
                inputProps={{
                  ...register('name', {
                    required: 'Please enter a valid artist name',
                  }),
                  defaultValue: artist?.name,
                  autoComplete: 'artist-name',
                }}
              />
            </div>
            {formState.errors.root && (
              <p>Error saving changes. Please try again later.</p>
            )}
            <div className="controls">
              <Button
                className="clear"
                disabled={false}
                category="secondary"
                text="Clear Form"
                type="reset"
              />
              <Button
                className="login"
                disabled={submitDisabled}
                text="Save"
                type="submit"
              />
            </div>
          </form>
        </DialogLayout>
      </dialog>
    </ProtectedRoute>
  );
};
