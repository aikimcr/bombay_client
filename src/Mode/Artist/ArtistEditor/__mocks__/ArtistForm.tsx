import React from 'react';
import { ComponentMock } from '../../../../testHelpers/mocks/ComponentMock';
import { ArtistModel } from '../../../../Model/ArtistModel';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const ArtistForm = ({
  artist,
}: {
  artist: ArtistModel;
  children: JSX.Element | JSX.Element[];
}) => {
  return (
    <ComponentMock
      name="Artist"
      classname="mock-artist-form"
      testId="mock-artist-form"
    >
      <p>{artist.id}</p>
      <p>{artist.name}</p>
    </ComponentMock>
  );
};
