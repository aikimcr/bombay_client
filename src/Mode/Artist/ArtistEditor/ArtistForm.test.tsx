import React, { useRef } from 'react';
import { act, fireEvent, render, screen } from '@testing-library/react';
import { http, HttpResponse } from 'msw';
import { setupServer } from 'msw/node';
import {
  setupArtistModelFetch,
  TestArtistCollectionURL,
} from '../../../Model/testing';
import { mockIntersectionObserver } from '../../../Hooks/testing';
import { ContextChanger } from '../../../testHelpers';
import { ArtistForm } from './ArtistForm';
import { ArtistData, ArtistModel } from '../../../Model/ArtistModel';

const fetchItems: Record<string, ReturnType<typeof setupArtistModelFetch>> = {};
const putItems: ReturnType<typeof setupArtistModelFetch>[] = [];
const postItems: ReturnType<typeof setupArtistModelFetch>[] = [];

const handlers = [
  http.put(`${TestArtistCollectionURL}/:id`, () => {
    return HttpResponse.json(putItems.shift()[1]);
  }),
  http.post(TestArtistCollectionURL, () => {
    return HttpResponse.json(postItems.shift()[1]);
  }),
];

const server = setupServer(...handlers);

beforeAll(() => {
  while (Object.keys(fetchItems).length <= 20) {
    const [getPromise, fetchDef] = setupArtistModelFetch();
    fetchItems[fetchDef.id.toString()] = [getPromise, fetchDef];
  }
});

afterAll(() => server.close());

beforeEach(() => {
  window.IntersectionObserver = mockIntersectionObserver;
  server.listen();
});
afterEach(() => {
  server.resetHandlers();
  delete window.IntersectionObserver;
  mockIntersectionObserver._clear();
});

describe('Test Artist editor form', () => {
  it('should match snapshot', async () => {
    const mockRef: ReturnType<typeof useRef<HTMLDialogElement>> = {
      current: null,
    };
    const [_getPromise, artistDef] = setupArtistModelFetch();
    const artist = new ArtistModel({
      data: artistDef,
    });

    const { asFragment } = render(
      <ContextChanger initialLoggedIn={true}>
        <ArtistForm artist={artist} ref={mockRef} />
      </ContextChanger>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should show the current artist name as the default value', async () => {
    const mockRef: ReturnType<typeof useRef<HTMLDialogElement>> = {
      current: null,
    };
    const [_getPromise, artistDef] = setupArtistModelFetch();
    const artist = new ArtistModel({
      data: artistDef,
    });

    render(
      <ContextChanger initialLoggedIn={true}>
        <ArtistForm artist={artist} ref={mockRef} />
      </ContextChanger>,
    );

    const form = screen.getByTestId('artist-form-form');
    const input = screen.getByLabelText('Artist Name');

    expect(form).toBeInTheDocument();
    expect(input).toBeInTheDocument();
    expect(form).toContain(input);
    expect(input).toHaveValue(artist.name);
  });

  it('should show the value as empty if no artist is specified', () => {
    const mockRef: ReturnType<typeof useRef<HTMLDialogElement>> = {
      current: null,
    };

    render(
      <ContextChanger initialLoggedIn={true}>
        <ArtistForm ref={mockRef} />
      </ContextChanger>,
    );

    const form = screen.getByTestId('artist-form-form');
    const input = screen.getByLabelText('Artist Name');

    expect(form).toBeInTheDocument();
    expect(input).toBeInTheDocument();
    expect(form).toContain(input);
    expect(input).toHaveValue('');
  });

  it('should save changes to the artist name', async () => {
    const mockRef: ReturnType<typeof useRef<HTMLDialogElement>> = {
      current: null,
    };
    const [getPromise, artistDef] = setupArtistModelFetch();
    const artist = ArtistModel.from<ArtistData, ArtistModel>(artistDef, {
      keepId: true,
    });

    expect(artist.id).toEqual(artistDef.id);

    render(
      <ContextChanger initialLoggedIn={true}>
        <ArtistForm artist={artist} ref={mockRef} />
      </ContextChanger>,
    );

    expect(mockRef.current).toBeDefined();
    const mockClosed = jest.fn();
    mockRef.current.close = mockClosed;

    const input = screen.getByLabelText('Artist Name');
    const submitButton = screen.getByText('Save');

    putItems.push([
      getPromise,
      {
        ...artistDef,
        name: 'xyzzy',
      },
    ]);

    await act(async () => {
      fireEvent.change(input, { target: { value: 'xyzzy' } });
    });

    expect(artist.id).toEqual(artistDef.id);
    expect(input).not.toHaveValue(artistDef.name);
    expect(input).toHaveValue('xyzzy');
    expect(artist.name).toEqual(artistDef.name);
    expect(mockClosed).not.toHaveBeenCalled();

    await act(async () => {
      fireEvent.click(submitButton);
    });

    expect(artist.id).toEqual(artistDef.id);
    expect(input).toHaveValue('xyzzy');
    expect(artist.name).toEqual('xyzzy');
    expect(mockClosed).toHaveBeenCalled();
  });
});

it('should save the new artist', async () => {
  const mockRef: ReturnType<typeof useRef<HTMLDialogElement>> = {
    current: null,
  };
  const [getPromise, artistDef] = setupArtistModelFetch();
  const artist = ArtistModel.from<ArtistData, ArtistModel>(artistDef);

  expect(artist.id).not.toBeDefined();

  render(
    <ContextChanger initialLoggedIn={true}>
      <ArtistForm artist={artist} ref={mockRef} />
    </ContextChanger>,
  );

  expect(mockRef.current).toBeDefined();
  const mockClosed = jest.fn();
  mockRef.current.close = mockClosed;

  const input = screen.getByLabelText('Artist Name');
  const submitButton = screen.getByText('Save');

  postItems.push([
    getPromise,
    {
      ...artistDef,
      name: 'xyzzy',
    },
  ]);

  await act(async () => {
    fireEvent.change(input, { target: { value: 'xyzzy' } });
  });

  expect(artist.id).not.toBeDefined();
  expect(input).not.toHaveValue(artistDef.name);
  expect(input).toHaveValue('xyzzy');
  expect(artist.name).toEqual(artistDef.name);
  expect(mockClosed).not.toHaveBeenCalled();

  await act(async () => {
    fireEvent.click(submitButton);
  });

  expect(artist.id).toEqual(artistDef.id);
  expect(input).toHaveValue('xyzzy');
  expect(artist.name).toEqual('xyzzy');
  expect(mockClosed).toHaveBeenCalled();
});
