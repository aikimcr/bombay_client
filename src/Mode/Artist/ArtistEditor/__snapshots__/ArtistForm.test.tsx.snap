// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Test Artist editor form should match snapshot 1`] = `
<DocumentFragment>
  <dialog
    class="artist-form"
    closedby="any"
  >
    <div
      class="dialog-layout"
    >
      <header
        class="dialog-layout__header"
      >
        <h3>
          Artist Editor
        </h3>
        <button
          class="close-button"
          data-testid="close-button"
          type="button"
        >
          <i
            class="fa-regular fa-circle-xmark"
          />
        </button>
      </header>
      <div
        class="dialog-layout__content"
      >
        <form
          class="artist-form__form"
          data-testid="artist-form-form"
        >
          <div
            class="info"
          >
            <div
              class="labeled-input"
            >
              <label
                class="labeled-input__label"
                for="name"
              >
                Artist Name
              </label>
              <input
                autocomplete="artist-name"
                class="text-input undefined"
                id="name"
                name="name"
                value="Spencer Davis Group"
              />
            </div>
          </div>
          <div
            class="controls"
          >
            <button
              class="clear button button__secondary"
              type="reset"
            >
              Clear Form
            </button>
            <button
              class="login button button__primary"
              disabled=""
              type="submit"
            >
              Save
            </button>
          </div>
        </form>
      </div>
    </div>
  </dialog>
  <button
    class="change-test-login"
    data-testid="change-test-login"
  >
    Change Login
  </button>
</DocumentFragment>
`;
